File: web/src/types/cash.ts
// Cash types - mirrors internal/domain/cash*.go

import type { Branch } from './branch'
import type { User } from './user'

export type CashSessionStatus = 'open' | 'closed'
export type CashMovementType = 'income' | 'expense'

export const CASH_SESSION_STATUSES: { value: CashSessionStatus; label: string; color: string }[] = [
  { value: 'open', label: 'Abierta', color: 'green' },
  { value: 'closed', label: 'Cerrada', color: 'gray' },
]

export const CASH_MOVEMENT_TYPES: { value: CashMovementType; label: string; color: string }[] = [
  { value: 'income', label: 'Ingreso', color: 'green' },
  { value: 'expense', label: 'Egreso', color: 'red' },
]

export interface CashRegister {
  id: number
  branch_id: number
  name: string
  code: string
  description?: string
  is_active: boolean
  created_at: string
  updated_at: string
  branch?: Branch
}

export interface CashSession {
  id: number
  cash_register_id: number
  branch_id: number
  user_id: number
  opening_amount: number
  closing_amount?: number
  expected_amount?: number
  difference?: number
  status: CashSessionStatus
  opened_at: string
  closed_at?: string
  opening_notes?: string
  closing_notes?: string
  closed_by?: number
  created_at: string
  updated_at: string
  register?: CashRegister
  branch?: Branch
  user?: User
  movements?: CashMovement[]
}

export interface CashMovement {
  id: number
  branch_id: number
  session_id: number
  movement_type: CashMovementType
  amount: number
  payment_method: string
  reference_type?: string
  reference_id?: number
  description: string
  balance_after: number
  created_by?: number
  created_at: string
  cash_session?: CashSession
  branch?: Branch
}

export interface CreateCashRegisterInput {
  branch_id: number
  name: string
  description?: string
}

export interface UpdateCashRegisterInput {
  name?: string
  description?: string
  is_active?: boolean
}

export interface OpenCashSessionInput {
  cash_register_id: number
  opening_amount: number
  opening_notes?: string
}

export interface CloseCashSessionInput {
  closing_amount: number
  closing_notes?: string
}

export interface CreateCashMovementInput {
  branch_id: number
  session_id: number
  movement_type: CashMovementType
  amount: number
  payment_method: string
  reference_type?: string
  reference_id?: number
  description: string
}

export interface CashSessionSummary {
  session_id: number
  opening_amount: number
  total_income: number
  total_expense: number
  expected_balance: number
  closing_amount?: number
  difference?: number
  movements_count: number
  movements_by_type: {
    type: CashMovementType
    count: number
    total: number
  }[]
}

export interface CashSessionListParams {
  page?: number
  per_page?: number
  order_by?: string
  order?: 'asc' | 'desc'
  register_id?: number
  branch_id?: number
  user_id?: number
  status?: CashSessionStatus
  date_from?: string
  date_to?: string
}

File: internal/domain/cash.go
package domain

import (
	"time"
)

// CashSessionStatus represents the status of a cash session
type CashSessionStatus string

const (
	CashSessionStatusOpen   CashSessionStatus = "open"
	CashSessionStatusClosed CashSessionStatus = "closed"
)

// CashMovementType represents the type of cash movement
type CashMovementType string

const (
	CashMovementTypeIncome  CashMovementType = "income"
	CashMovementTypeExpense CashMovementType = "expense"
)

// CashRegister represents a physical cash register
type CashRegister struct {
	ID          int64   `json:"id"`
	BranchID    int64   `json:"branch_id"`
	Name        string  `json:"name"`
	Code        string  `json:"code"`
	Description *string `json:"description,omitempty"`
	IsActive    bool    `json:"is_active"`

	// Timestamps
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`

	// Relations
	Branch *Branch `json:"branch,omitempty"`
}

// TableName returns the database table name
func (CashRegister) TableName() string {
	return "cash_registers"
}

// CashSession represents a cash register session
type CashSession struct {
	ID               int64 `json:"id"`
	BranchID         int64 `json:"branch_id"`
	CashRegisterID   int64 `json:"cash_register_id"`
	UserID           int64 `json:"user_id"`

	// Amounts
	OpeningAmount  float64  `json:"opening_amount"`
	ClosingAmount  *float64 `json:"closing_amount,omitempty"`
	ExpectedAmount *float64 `json:"expected_amount,omitempty"`
	Difference     *float64 `json:"difference,omitempty"`

	// Status
	Status CashSessionStatus `json:"status"`

	// Timestamps
	OpenedAt time.Time  `json:"opened_at"`
	ClosedAt *time.Time `json:"closed_at,omitempty"`

	// Notes
	OpeningNotes *string `json:"opening_notes,omitempty"`
	ClosingNotes *string `json:"closing_notes,omitempty"`

	// Audit
	ClosedBy *int64 `json:"closed_by,omitempty"`

	// Timestamps
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`

	// Relations
	CashRegister *CashRegister   `json:"register,omitempty"`
	User         *User           `json:"user,omitempty"`
	Branch       *Branch         `json:"branch,omitempty"`
	Movements    []*CashMovement `json:"movements,omitempty"`
}

// TableName returns the database table name
func (CashSession) TableName() string {
	return "cash_sessions"
}

// IsOpen checks if the session is open
func (cs *CashSession) IsOpen() bool {
	return cs.Status == CashSessionStatusOpen
}

// CashMovement represents a cash movement in a session
type CashMovement struct {
	ID        int64 `json:"id"`
	BranchID  int64 `json:"branch_id"`
	SessionID int64 `json:"session_id"`

	// Movement details
	MovementType  CashMovementType `json:"movement_type"`
	Amount        float64          `json:"amount"`
	PaymentMethod PaymentMethod    `json:"payment_method"`

	// Reference to related entity
	ReferenceType *string `json:"reference_type,omitempty"` // loan, payment, sale
	ReferenceID   *int64  `json:"reference_id,omitempty"`

	// Description
	Description string `json:"description"`

	// Balance
	BalanceAfter float64 `json:"balance_after"`

	// Audit
	CreatedBy int64     `json:"created_by,omitempty"`
	CreatedAt time.Time `json:"created_at"`

	// Relations
	CashSession *CashSession `json:"cash_session,omitempty"`
	Branch      *Branch      `json:"branch,omitempty"`
}

// TableName returns the database table name
func (CashMovement) TableName() string {
	return "cash_movements"
}

// IsIncome checks if the movement is an income
func (cm *CashMovement) IsIncome() bool {
	return cm.MovementType == CashMovementTypeIncome
}

// IsExpense checks if the movement is an expense
func (cm *CashMovement) IsExpense() bool {
	return cm.MovementType == CashMovementTypeExpense
}
